src/compute-client/src/as_of_selection.rs

```diff
@@ -0,0 +1,699 @@
+        // so we use a fixpoint loop here.
+        fixpoint(|changed| {
+            self.propagate_bounds_upstream_inner(bound_type, constraint_type, changed);
+
+            // Propagating `lower` bounds upstream might break `AsOfBounds` invariant (1), so we
+            // need to restore it.
+            if bound_type == BoundType::Lower {
+                self.propagate_bounds_downstream_inner(
+                    BoundType::Lower,
+                    ConstraintType::Hard,
+                    changed,
+                );
+            }
+        });
+    }
+
+    fn propagate_bounds_upstream_inner(
+        &self,
+        bound_type: BoundType,
+        constraint_type: ConstraintType,
+        changed: &mut bool,
+    ) {
+        for (id, collection) in self.collections.iter().rev() {
+            let bounds = collection.bounds.borrow();
+            for input_id in &collection.compute_inputs {
+                let constraint = Constraint {
+                    type_: constraint_type,
+                    bound_type,
+                    frontier: bounds.get(bound_type),
+                    reason: &format!("downstream {id} {bound_type} as-of bound"),
+                };
+                *changed |= self.apply_constraint(*input_id, constraint);
+            }
+        }
+    }
+
+    /// Selects the "best" as-of for the identified collection, based on its currently known
+    /// bounds.
+    ///
+    /// We simply use the upper bound here, to maximize the chances of compute reconciliation
+    /// succeeding. Choosing the latest possible as-of also minimizes the amount of work the
+    /// dataflow has to spend processing historical data from its sources.
+    fn best_as_of(&self, id: GlobalId) -> Antichain<T> {
+        if let Some(collection) = self.collections.get(&id) {
+            let bounds = collection.bounds.borrow();
+            bounds.upper.clone()
+        } else {
+            Antichain::new()
+        }
+    }
+
+    /// Removes collections that sink into sealed persist shards from the context.
+    ///
+    /// The dataflows of these collections will get an empty default as-of assigned at the end of
+    /// the as-of selection process, ensuring that they won't get installed unnecessarily.
+    ///
+    /// Note that it is valid to remove these collections from consideration because they don't
+    /// impose as-of constraints on other compute collections.
+    fn prune_sealed_persist_sinks(&mut self) {
+        self.collections.retain(|id, _| {
+            self.storage_collections
+                .collection_frontiers(*id)
+                .map_or(true, |f| !f.write_frontier.is_empty())
+        });
+    }
+}
+
+/// Runs `step` in a loop until it stops reporting changes.
+fn fixpoint(mut step: impl FnMut(&mut bool)) {
+    loop {
+        let mut changed = false;
+        step(&mut changed);
+        if !changed {
+            break;
+        }
+    }
+}
+
+/// Step back the given frontier, if possible.
```

Comment:
This method is only used in a single place and does exactly what the single place needs: saturate instead of erroring. I expanded the doc comment to include that.